### DISCUSSION TITLE
Workflows vs actions: which permission is correct?

### DISCUSSION THREAD

**Discussion by zkamvar** [community/community#26711] [Actions]
<p>I want to have a workflow in my repository that periodically updates other workflows in the repository on a semi-weekly basis, so I’m trying to <a href="https://github.blog/changelog/2021-04-20-github-actions-control-permissions-for-github_token/" rel="noopener nofollow ugc">use job control permissions for GITHUB_TOKEN</a>.</p>
<p>I have a MWE here: <a href="https://github.com/zkamvar/cautious-octo-chainsaw/actions/workflows/update.yml" class="inline-onebox" rel="noopener nofollow ugc">Actions · zkamvar/cautious-octo-chainsaw · GitHub</a></p>
<p>I first tried to use <code>actions: write</code> because that was documented in the blog post, but when I do that, <a href="https://github.com/zkamvar/cautious-octo-chainsaw/runs/2496735624?check_suite_focus=true#step:5:76" rel="noopener nofollow ugc">I get an error message saying that I need the <code>workflow</code> permission</a>:</p>
<pre><code class="lang-auto">! [remote rejected] HEAD -&gt; update-workflow-2021-05-04-00-49 (refusing to allow a GitHub App to create or update workflow `.github/workflows/remove-branch.yaml` without `workflows` permission)
</code></pre>
<p>I then tried to use the <code>workflows: write</code> permission and then <a href="https://github.com/zkamvar/cautious-octo-chainsaw/actions/runs/808547704/workflow" rel="noopener nofollow ugc">I get a new error</a> that says the workflow is not valid because <code>workflows</code> is not a recognized permissions value.</p>
<pre><code class="lang-auto">The workflow is not valid. 

github/workflows/update.yml (Line: 18, Col: 7): Unexpected value 'workflows'
</code></pre>
<p>How can I write a workflow that has permissions to make pull requests updating other workflows?</p>

**Comment by otter-computer** [comment, marked as answer]
<p>The <code>permissions</code> block for the <code>GITHUB_TOKEN</code> allows you to set the <a href="https://docs.github.com/en/rest/reference/permissions-required-for-github-apps">GitHub App permissions</a> for the token. The <code>actions</code> permission only grants you <a href="https://docs.github.com/en/rest/reference/permissions-required-for-github-apps#permission-on-actions">some permissions for accessing Actions APIs</a>, but not access to updating workflow files. This is to prevent cases of Actions triggering an infinite loop.</p>
<p>If you want to create a workflow that can update workflow files then you’ll need to provide a <a href="https://docs.github.com/en/developers/apps/about-apps#personal-access-tokens">Personal Access Token</a> as a secret that <a href="https://docs.github.com/en/developers/apps/scopes-for-oauth-apps">grants the <code>workflow</code> scope</a> and use that in place of the provided token.</p>

**Comment by rcowsill** [comment]
<p>This restriction causes problems for “rebase” actions such as <a href="https://github.com/cirrus-actions/rebase" rel="noopener nofollow ugc">cirrus-actions/rebase</a>. If the upstream repo has workflow changes then rebases of forked PRs get blocked.</p>
<p>Using a PAT to work around this is risky, as it gives a long-lived credential to a workflow operating on untrusted data.</p>
<p>Is there any way to support this workflow while keeping the benefits of the short-lived github token?</p>

**Comment by ryan-lane** [comment]
<p>I’d really like github to reassess this issue, and maybe find a way to not trigger infinite loops. Very basic things like needing to auto-update a tag, or do rebases, is causing actions to be unusable without a PAT, if the changes include any changes to a workflow.</p>
<p>We shouldn’t be encouraging people to put PATs into their environment across large sets of repos in their organization, because unlike the tokens in the actions, the PATs have much wider access rights, and it will allow repos to modify other repos through their actions.</p>
<p>As an example, I want to have a common action for all repos that rebases PRs. Now I need to add a PAT. The only way to sanely do this is to use a PAT that has access to all the repos. Now a developer who only had rights to update a single repo can update every repo that the PAT allows, through their actions.</p>
<p>Think about organizations who are using terraform, triggered through github actions. Even running a plan could have side-effects that escalate privileges (and not everyone knows that), so using a PAT like this may give someone with only access to one repo much, much higher privileges.</p>
<p>Also, couldn’t you have the same infinite loop situation if the action is changing an action via a PAT? I’m not sure what this is doing other than lowering everyone’s security.</p>

**Reply by bplessis-swi** [reply]
I am currently implementing an Github App replacement of a PAT for this exact use case ( based on https://dev.to/dtinth/authenticating-as-a-github-app-in-a-github-actions-workflow-27co  )

However the permission availables for a Github App (which are the same as fine-grained PAT) don't seem to prevent this loop issue.

I have allowed for `Content` and `Workflows` to Read/Write but left `Actions` to none yet the workflow change do trigger an inifinite loop

**Reply by SvenStaehs** [reply]
Just to add my use-case: I uploaded a repo from github.com to our GHES instance. There is no support for forking from cloud into GHES, so I had to do this manually.  
To at least make continuous maintenance a bit easier, I set up a workflow to push upstream's main to a branch in my GHES repo every night, but ran into the above error.

So looking forward to a solution for this.  
I don't really understand the issue with infinite loop. Do you mean a workflow triggered on push modifying its own branch, thus triggering another push? But then how does using a PAT prevent this?
At some point you have to be content telling people about the danger and risk them making mistakes, otherwise you have no product. You can still mitigate against the possible effect, e.g. rate-restrict workflows triggered by a push that was performed by an action. If people are using PATs on the other hand, you have no chance of detecting that, because it looks like a person did the push...
